lib/app_profiler/server.rb

```diff
@@ -0,0 +1,292 @@
+        @semaphore.synchronize do
+          return false if @profile_running
+
+          @profile_running = true
+        end
+      end
+
+      def stop_running
+        @semaphore.synchronize { @profile_running = false }
+      end
+    end
+
+    # This is a minimal, non-compliant "HTTP" server.
+    # It will create an extremely minimal rack environment hash and hand it off
+    # to our application to process
+    class ProfileServer
+      PROFILER_TEMPFILE_PATH = "/tmp/app_profiler" # for tempfile that indicates port in filename or unix sockets
+
+      class ProfileServerTransport
+        attr_reader :socket
+
+        def client
+          raise("'client' method not implemented for transport")
+        end
+
+        def stop
+          raise("'stop' method not implemented for transport")
+        end
+      end
+
+      class ProfileServerTransportUNIX < ProfileServerTransport
+        def initialize
+          super
+          FileUtils.mkdir_p(PROFILER_TEMPFILE_PATH)
+          @socket_file = File.join(PROFILER_TEMPFILE_PATH, "app-profiler-#{Process.pid}.sock")
+          File.unlink(@socket_file) if File.exist?(@socket_file) && File.socket?(@socket_file)
+          @socket = UNIXServer.new(@socket_file)
+        end
+
+        def client
+          UNIXSocket.new(@socket_file)
+        end
+
+        def stop
+          File.unlink(@socket_file) if File.exist?(@socket_file) && File.socket?(@socket_file)
+          @socket.close
+        end
+      end
+
+      class ProfileServerTransportTCP < ProfileServerTransport
+        SERVER_ADDRESS = "127.0.0.1" # it is ONLY safe to run this bound to localhost
+
+        def initialize(port = 0)
+          super()
+          FileUtils.mkdir_p(PROFILER_TEMPFILE_PATH)
+          @socket = TCPServer.new(SERVER_ADDRESS, port)
+          @port = @socket.addr[1]
+          @port_file = Tempfile.new("profileserver-#{Process.pid}-port-#{@port}-", PROFILER_TEMPFILE_PATH)
+        end
+
+        def client
+          TCPSocket.new(SERVER_ADDRESS, @port)
+        end
+
+        def stop
+          @port_file.unlink
+          @socket.close
+        end
+      end
+
+      def initialize(transport)
+        case transport
+        when TRANSPORT_UNIX
+          @transport = ProfileServer::ProfileServerTransportUNIX.new
+        when TRANSPORT_TCP
+          @transport = ProfileServer::ProfileServerTransportTCP.new(AppProfiler::Server.port)
+        else
+          raise "invalid transport #{transport}"
+        end
+
+        @listen_thread = nil
+
+        AppProfiler.logger.info(
+          "[AppProfiler::Server] listening on addr=#{@transport.socket.addr}"
+        )
+        at_exit { stop }
+      end
+
+      def client
+        @transport.client
+      end
+
+      def serve
+        return unless @listen_thread.nil?
+
+        app = ProfileApplication.new
+
+        @listen_thread = Thread.new do
+          loop do
+            Thread.new(@transport.socket.accept) do |session|
```

Comment:
This will spin up a bunch of threads in the loop. Is this intended, or should the loop be inside this block?